wgpu/src/lib.rs

```diff
@@ -1740,258 +1836,32 @@ impl Instance {
-    /// which has no lifetime requirements to `window` and doesn't require
-    /// [`Send`] or [`Sync`] (on non-Wasm targets). This makes it `unsafe`
-    /// instead and always returns a [`Surface<'static>`].
-    ///
-    /// See [`create_surface()`](Self::create_surface) for more details.
-    ///
-    /// # Safety
-    ///
-    /// - `raw_window_handle` must be a valid object to create a surface upon.
-    /// - `raw_window_handle` must remain valid until after the returned [`Surface`] is
-    ///   dropped.
-    pub unsafe fn create_surface_from_raw<W>(
-        &self,
-        window: &W,
-    ) -> Result<Surface<'static>, CreateSurfaceError>
-    where
-        W: HasWindowHandle + HasDisplayHandle,
-    {
-        let raw_display_handle = window
-            .display_handle()
-            .map_err(|e| CreateSurfaceError {
-                inner: CreateSurfaceErrorKind::RawHandle(e),
-            })?
-            .as_raw();
-        let raw_window_handle = window
-            .window_handle()
-            .map_err(|e| CreateSurfaceError {
-                inner: CreateSurfaceErrorKind::RawHandle(e),
-            })?
-            .as_raw();
-        let (id, data) = unsafe {
-            DynContext::instance_create_surface(
-                &*self.context,
-                raw_display_handle,
-                raw_window_handle,
-            )
-        }?;
-        Ok(Surface {
-            context: Arc::clone(&self.context),
-            _surface: None,
-            id,
-            data,
-            config: Mutex::new(None),
-        })
-    }
-
-    /// Creates a surface from `CoreAnimationLayer`.
-    ///
-    /// # Safety
-    ///
-    /// - layer must be a valid object to create a surface upon.
-    #[cfg(metal)]
-    pub unsafe fn create_surface_from_core_animation_layer(
-        &self,
-        layer: *mut std::ffi::c_void,
-    ) -> Surface<'static> {
-        let surface = unsafe {
-            self.context
-                .as_any()
-                .downcast_ref::<crate::backend::Context>()
-                .unwrap()
-                .create_surface_from_core_animation_layer(layer)
-        };
-        Surface {
-            context: Arc::clone(&self.context),
-            _surface: None,
-            id: ObjectId::from(surface.id()),
-            data: Box::new(surface),
-            config: Mutex::new(None),
-        }
-    }
-
-    /// Creates a surface from `IDCompositionVisual`.
+    /// Creates a new surface targeting a given window/canvas/surface/etc..
     ///
-    /// # Safety
+    /// See [`SurfaceTarget`] for what targets for surface are supported.
     ///
-    /// - visual must be a valid IDCompositionVisual to create a surface upon.
-    #[cfg(dx12)]
-    pub unsafe fn create_surface_from_visual(
+    /// Most commonly used are window handles (or provider of windows handles)
+    /// which can be passed directly as they're automatically converted to [`SurfaceTarget`].
+    pub fn create_surface<'window>(
```

Comment:
This isn't safe, unfortunately. 

I can very easily do

`instance.create_surface(SurfaceTarget::CoreAnimationLayer(std::ptr::null_mut()))` and bing bong boom segfault in safe code.

The "other" ones need to be unsafe somehow. Either by splitting this into two methods, one which accepts a raw-window-handle through ownership, one which accepts a raw variant.

Additionally I think having the raw variant in the unsafe bin is important, in case users have weird requirements where they can't hand us ownership of the window.